sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -112,6 +112,14 @@ object TypeCoercion {
         StructField(f1.name, dataType, nullable = f1.nullable || f2.nullable)
       }))
 
+    case (a1 @ ArrayType(et1, hasNull1), a2 @ ArrayType(et2, hasNull2)) if a1.sameType(a2) =>
+      findTightestCommonType(et1, et2).map(ArrayType(_, hasNull1 || hasNull2))
+
+    case (m1 @ MapType(kt1, vt1, hasNull1), m2 @ MapType(kt2, vt2, hasNull2)) if m1.sameType(m2) =>
+      val keyType = findTightestCommonType(kt1, kt2)
+      val valueType = findTightestCommonType(vt1, vt2)
```

Comment:
BTW, I think we should do the same thing in `findWiderTypeForTwo` to cover some corner cases such as decimal or string promotion within keys and values .. ? and seems https://github.com/apache/spark/pull/21100#discussion_r183032551 suggested the same thing ..?

Reply:
This is something we should figure out: why `findWiderTypeForTwo` only take care of array type? seems all complex types should be handled there, especially if it follows Hive's behavior.

Anyway it's orthogonal to `findTightestCommonType`, they are used for different operators.

Reply:
Yea, I was just wondering while reading it. However, doesn't that mean we don't do type widening for nested types in the same way? I was thinking we should do the same type widening for nested types too.

Reply:
I mean, I was thinking we should do that in both places in `findTightestCommonType` and `findWiderTypeForTwo`. Otherwise, the nested types in struct, map or array won't do, for example, decimal or string promotion.

Reply:
Sure, it's orthogonal. Yup, I was just wondering. I am okay to leave this out of this PR.

Reply:
We also need to look into `findTightestCommonType`. Currently we are just very conservative and only allow nullability change for complex types in `findTightestCommonType`. We need to take a look at other systems and see what they do.

Reply:
I agree with that given past discussions - I didn't mean we should change something now .. but was just wondering.